dependencies: upgrade ginkgo to v2

dolmen opened the pull request:
Upgrade ginko to v2

## Description

1. `go get github.com/onsi/ginkgo/v2@latest`
2. `go get github.com/onso/gomega@latest`
3. Fix `reference-gen/*_test.go` to use `ginkgo/v2`
4. `go mod tidy`
4. `go get k8s.io/klog/v2@latest` # fix nil logger errors
5. `go mod tidy`
6. `go mod edit -go=1.23` # instead of 1.23.0
7. `go mod tidy`

## Motivation and Context

Dependencies are outdated. Those dependencies are affecting the go.mod of `oauth2-proxy`. Among others, they bring a dependency on `gopkg.in/yaml.v2`.

## How Has This Been Tested?

`go test ./...`

dolmen committed:
deps: upgrade ginkgo to v2

Upgrade ginko to v2:
1. go get github.com/onsi/ginkgo/v2@latest
2. go get github.com/onso/gomega@latest
3. Fix reference-gen/*_test.go to use ginkgo/v2
4. go mod tidy
4. go get k8s.io/klog/v2@latest # fix nil logger errors
5. go mod tidy
6. go mod edit -go=1.23 # instead of 1.23.0
7. go mod tidy

dolmen commented:
Cc: @tuunit This is somewhat related to oauth2-proxy/oauth2-proxy#2982 : this is also a dependency cleanup.

tuunit commented on reference-gen/go.mod:
```diff
@@ -1,25 +1,27 @@
-go 1.22
+go 1.23.0
+
+toolchain go1.24.1
```
Switching to 1.24 is a little too soon for my liking

dolmen replied:
1. This is switching the minimum Go version to 1.23, not 1.24, and this change comes from one of the dependencies
2. This project is just a tool for code generation for `oauth2-proxy`. Only developers of `oauth2-proxy` are supposed to run this tool. This change will not affect end-users of `oauth2-proxy`, especially thanks to oauth2-proxy/oauth2-proxy#2982

I can remove the `toolchain` line if you prefer, but I would like to understand your concerns about running the go:generate steps of oauth2-proxy with a tool built with the most recent stable release of Go.

tuunit replied:
First of all, golangci-lint isn't yet compatible with your change:
https://github.com/oauth2-proxy/tools/actions/runs/13786253316/job/39002341332

Go 1.24 was just released a couple of weeks ago and if you have ever worked with larger projects that other depend upon you never want to update to a new major version immediately after it was released.

A lot of people have custom builds and forks of oauth2-proxy and we will definitely cause some unforeseen side effects if we upgrade to soon. As you can tell with golangci-lint.

tuunit replied:
Doesn't really matter too much for the tool repo but I will keep off upgrading oauth2-proxy to v1.24 any time soon

dolmen replied:
I'll just remove the `toolchain` line and see how it goes...

dolmen replied:
```console
$ go mod edit -toolchain=go1.23.7
```

dolmen committed:
reference-gen: downgrade Go toolchain to go1.23

The "toolchain go1.24.1" had been automatically added in go.mod. Let's downgrade
to match the "go" directive.

go mod edit -toolchain=go1.23.7

dolmen commented:
@tuunit reference-gen/go.mod now requires only Go 1.23.
